File: Blind150/7.Tree/Python/kth_smallest_element_BST.py
# https://leetcode.com/problems/kth-smallest-element-in-a-bst/description/
from typing import Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:

        elms = []

        def inorder(root):

            nonlocal elms

            if not root:
                return

            inorder(root.left)

            if len(elms) == k:
                return

            elms.append(root.val)

            inorder(root.right)

        inorder(root)

        return elms[-1]

File: Blind150/7.Tree/Go/kth_smallest_element_BST.go
// https://leetcode.com/problems/kth-smallest-element-in-a-bst/description/
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func kthSmallest(root *TreeNode, k int) int {

	var elems []int

	var inorder func(root *TreeNode)

	inorder = func(root *TreeNode) {

		if root == nil {
			return
		}

		inorder(root.Left)

		if len(elems) == k {
			return
		}

		elems = append(elems, root.Val)

		inorder(root.Right)
	}

	inorder(root)

	return elems[len(elems)-1]

}
